analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs

```diff
@@ -48,15 +49,80 @@ public abstract class EmptyCollectionsShouldNotBeEnumeratedBase : SymbolicRuleCh
         KnownType.System_Collections_Generic_Dictionary_TKey_TValue,
         KnownType.System_Collections_Generic_IDictionary_TKey_TValue,
         KnownType.System_Collections_Immutable_IImmutableDictionary_TKey_TValue,
+        KnownType.System_Collections_Generic_IReadOnlyCollection_T,
+        KnownType.System_Collections_Generic_IReadOnlyList_T,
     }.ToImmutableArray();
 
     protected static readonly HashSet<string> RaisingMethods = new()
     {
-        nameof(IEnumerable<int>.GetEnumerator),
+        "TryDequeue",
+        "TryPeek",
+        "TryPop",
+        nameof(Array.Clone),
+        nameof(Array.Copy),
+        nameof(Array.GetLength),
+        nameof(Array.GetLongLength),
+        nameof(Array.GetLowerBound),
+        nameof(Array.GetUpperBound),
+        nameof(Array.GetValue),
+        nameof(Array.Initialize),
+        nameof(Array.SetValue),
+        nameof(Dictionary<int, int>.ContainsKey),
+        nameof(Dictionary<int, int>.ContainsValue),
+        nameof(Dictionary<int, int>.TryGetValue),
+        nameof(Enumerable.Aggregate),
+        nameof(Enumerable.All),
+        nameof(Enumerable.Any),
+        nameof(Enumerable.Average),
+        nameof(Enumerable.Cast),
+        nameof(Enumerable.Contains),
+        nameof(Enumerable.Count),
```

Comment:
There's `Count` here, and it doesn't raise in UTs. Something is wrong. 
I didn't check every single one of them, but we need UT for every single one

Reply:
Count from `KnownType.System_Linq_Enumerable` has a special treatment (first nested `if` in ProcessInvocation) to learn the constraint from it, I might add a bullet point here to take a further look into it and improve it.

Reply:
That's independent handling for NumberConstraint. It should not affect raising an issue on it.

Reply:
Ah, the problem is the return.
So 
```
var state = context.State;`
...use it everywhere
..and inside the Count
   state = State.SetOperationConstraint(context.Operation, constraint);